_posts/2020-06-27-lsp-explained.md

```diff
@@ -0,0 +1,183 @@
+---
+title: The Liskov Substitution Principle Explained
+categories: [craft]
+date: 2020-06-27 05:00:00 +1100
+modified: 2020-06-27 05:00:00 +1100
+author: saajan
+excerpt: "LSP is a very useful idea both when developing new applications and modifying existing ones. This article explains what it is, why it's important and how to use it."
+image:
+  auto: 0071-disk
+---
+
+This article gives a quick intro to the Liskov Substitution Principle (LSP), why it's important, and how we can use it to validate our object-oriented designs. We'll also see some examples and learn how to correctly identify violations of LSP.
+
+{% include github-project.html url="https://github.com/thombergs/code-examples/tree/master/solid" %}
+
+# What is LSP?
+
+In simple terms, **LSP states that in an object-oriented program, if we substitute a superclass object reference with an object of *any* of its subclasses, the program should not break.** 
+
+Essentially, if we had a method call that takes a superclass object reference as a parameter, like below:
+
+```java
+doSomething(superclassObjectReference);
```

Comment:
Can you include the whole class around this line of code? That would make it more clear.

Reply:
Ok, I thought keeping just the essential line would make it easier. I have added the whole class and also rephrased the line starting with "Essentially...". Please see if this is better.